packages/core/src/scp-cf/destination-accessor.ts

```diff
@@ -482,19 +285,346 @@ const emptyDestinationByType: DestinationsByType = {
   subaccount: []
 };
 
-function subscriberDestinationIsSelected(
-  selectionStrategy: DestinationSelectionStrategy
-): boolean {
-  return (
-    selectionStrategy === subscriberFirst ||
-    selectionStrategy === alwaysSubscriber
-  );
-}
-
 function emptyDestinationsByType(
   destinationByType: DestinationsByType
 ): boolean {
   return (
-    !destinationByType.instance.length && !destinationByType.instance.length
+    !destinationByType.instance.length && !destinationByType.subaccount.length
   );
 }
+
+class DestinationAccessor {
+  public static async getDestinationFromDestinationService(
+    name: string,
+    options: DestinationOptions & { iss?: string }
+  ): Promise<Destination | undefined> {
+    const decodedUserJwt = await DestinationAccessor.getDecodedUserJwt(options);
+    const providerToken = await DestinationAccessor.getProviderToken(options);
+    const da = new DestinationAccessor(
+      name,
+      options,
+      decodedUserJwt,
+      providerToken
+    );
+
+    return (
+      da.trySubscriberCache() ??
+      (await da.tryProviderCache()) ??
+      (await da.processDestinationResult(await da.tryForOAuth())) ??
+      (await da.processDestinationResult(await da.tryForCert())) ??
+      (await da.processDestinationResult(await da.tryService())) ??
```

Comment:
two observations:
1. I think you're calling `tryService` in all three functions. given the cache, that shouldn't be an issue, but at least in my mental model I would've first called `tryService` explicitly and then passed it through these functions. But that's rather minor, and you could probably also make a case for why it makes sense to do it the way you've done it.
2. What bugs me more is that if you call `processDestinationResult` in every case, then using some kind of pipeline-thingy would make more sense here. Maybe not exactly that fp pipeline thingy, but write in a way that you don't have to explicitly call that function everytime. like:
```
const result = x ?? y ?? z;
processResult(result)
return result;
```

Reply:
I adjust it. Marika commented the same thing :-).